sql-plugin/src/main/scala/com/nvidia/spark/rapids/spill/SpillFramework.scala

```diff
@@ -1597,59 +1591,72 @@ object SpillFramework extends Logging {
- * buffer is in use will block at `acquire`. Calls to `release` notify the blocked
- * threads, and they will check to see if they can acquire.
+ * The bounce buffer is acquired from a BounceBufferPool, so any calls to
+ * BounceBufferPool.nextBuffer will block if the pool has no available buffers.
  *
- * `close` is the interface to unacquire the bounce buffer.
+ * `close` restores a bounce buffer to the pool for other callers to use.
  *
- * `release` actually closes the underlying DeviceMemoryBuffer, and should be called
+ * `release` actually closes the underlying buffer, and should be called
  * once at the end of the lifetime of the executor.
  *
- * @param dmb - actual cudf DeviceMemoryBuffer that this class is protecting.
+ * @param buf - actual cudf DeviceMemoryBuffer that this class is protecting.
+ * @param pool - the pool to which this buffer belongs
  */
-private[spill] case class DeviceBounceBuffer(var dmb: DeviceMemoryBuffer) extends AutoCloseable {
-  private var acquired: Boolean = false
-  def acquire(): DeviceBounceBuffer = synchronized {
-    while (acquired) {
-      wait()
-    }
-    acquired = true
-    this
-  }
-
-  private def unaquire(): Unit = synchronized {
-    acquired = false
-    notifyAll()
-  }
-
+private[spill] class BounceBuffer[T <: AutoCloseable](
+                                                       var buf: T,
+                                                       private val pool: BounceBufferPool[T])
+  extends AutoCloseable {
   override def close(): Unit = {
-    unaquire()
+    pool.returnBuffer(this)
   }
 
-  def release(): Unit = synchronized {
-    if (acquired) {
-      throw new IllegalStateException(
-        "closing device buffer pool, but some bounce buffers are in use.")
-    }
-    if (dmb != null) {
-      dmb.close()
-      dmb = null
-    }
+  def release(): Unit = {
+    buf.close()
+    buf = null.asInstanceOf[T]
   }
 }
 
+
 /**
- * A bounce buffer pool with buffers of size `bufferSize`
+ * A bounce buffer pool with buffers of size `bufSize`
  *
- * This pool returns instances of `DeviceBounceBuffer`, that should
+ * This pool returns instances of `BounceBuffer[T]`, that should
  * be closed in order to be reused.
  *
  * Callers should synchronize before calling close on their `DeviceMemoryBuffer`s.
  */
-trait DeviceBounceBufferPool extends AutoCloseable {
-  def bufferSize: Long
-  def nextBuffer(): DeviceBounceBuffer
+class BounceBufferPool[T <: AutoCloseable](private val bufSize: Long,
+                                           private val bbCount: Long,
+                                           private val allocator: Long => T)
+  extends AutoCloseable with Logging {
+
+  private val pool = new LinkedBlockingQueue[BounceBuffer[T]]
+  for (_ <- 1L to bbCount) {
+    pool.offer(new BounceBuffer[T](allocator(bufSize), this))
+  }
+
+  def bufferSize: Long = bufSize
+  def nextBuffer(): BounceBuffer[T] = pool.take()
```

Comment:
@abellina Are you concerned about potential shutdown issues where this might return a BounceBuffer that has been released if close is called first? The code should be safe and not corrupt memory or anything. It is more of a question about how simple it is to debug this code.